spotify.go

```diff
@@ -124,6 +171,78 @@ func decodeError(r io.Reader) error {
 // authenticate, you can use `DefaultClient`.
 type Client struct {
 	http *http.Client
+
+	autoRetry     bool
+	retryDuration time.Duration
+}
+
+func (c *Client) SetAutoRetry(flag bool) {
+	c.autoRetry = flag
+}
+
+// executeOpt executes a non-GET request. `needsStatus` describes another code
+// that can represent success. Note that in all current usages of this function,
+// we need to still allow a 200 even if we'd also like to check for a second
+// success code.
+func (c *Client) executeOpt(req *http.Request, needsStatus int, result interface{}) error {
+	for {
+		resp, err := c.http.Do(req)
+		if err != nil {
+			return err
+		}
+
+		defer resp.Body.Close()
+
+		if resp.StatusCode == rateLimitExceededStatusCode && c.autoRetry {
+			time.Sleep(c.retryDuration)
+			continue
+		} else if resp.StatusCode != http.StatusOK && (needsStatus == 0 || resp.StatusCode != needsStatus) {
+			errorMessage := decodeError(c, resp)
+			return errorMessage
+		}
+
+		if result != nil {
+			if err := json.NewDecoder(resp.Body).Decode(result); err != nil {
+				return err
+			}
+		}
+
+		break
+	}
+
+	return nil
+}
+
+func (c *Client) execute(req *http.Request) error {
+	return c.executeOpt(req, 0, nil)
+}
+
+func (c *Client) get(url string, result interface{}) error {
+	for {
+		resp, err := c.http.Get(url)
+		if err != nil {
+			return err
+		}
+
+		defer resp.Body.Close()
+
+		if resp.StatusCode == rateLimitExceededStatusCode && c.autoRetry {
+			time.Sleep(c.retryDuration)
+			continue
+		} else if resp.StatusCode != http.StatusOK {
+			errorMessage := decodeError(c, resp)
```

Comment:
Nitpick, but `decodeError` returns an `Error`, not a message, so let's not use `errorMessage` for a variable name.

In fact, just `return decodeError(c, resp)` would work fine.